[Mission2. JPA] 3단계 - 질문 삭제하기 리팩터링

gregolee opened the pull request:
안녕하세요. 강현구 리뷰어님! 🙇
말씀하셨던 부분 반영하여 리뷰 재요청 드립니다!
항상 좋은 리뷰 👀 감사드리고 💯 👍 좋은 하루되시길 바랍니다!!! 🙇

gregolee committed:
docs: step03.md 추가

- markdown 추가 사항 커밋
- 요구사항 정리
   - step03.md

Resolve:
See Also: #134

gregolee committed:
chore: 강현구 리뷰어님께서 말씀하신 코멘트 반영

- [x] 0-5.리뷰어님의 리뷰를 반영한 코드로 수정
        - [x] 0-5-1.테스트 메서드 실행 순서가 서로 의존하게 작성하지 않기
            - [x] 0-5-1-1.`AnswerTest`
            - [x] 0-5-1-2.`AnswerRepositoryTest`
            - [x] 0-5-1-3.`DeleteHistoryTest`
            - [x] 0-5-1-4.`DeleteHistoryRepositoryTest`
            - [x] 0-5-1-5.`QuestionTest`
            - [x] 0-5-1-6.`QuestionRepositoryTest`
            - [x] 0-5-1-7.`UserTest`
            - [x] 0-5-1-8.`UserRepositoryTest`
        - [x] 0-5-2.클린코드 - 의미있게 구분하라. 그리고 테스트 코드는 이류 시민이 아니다.
            - [x] 0-5-2-1.`AnswerTest`
            - [x] 0-5-2-2.`AnswerRepositoryTest`
            - [x] 0-5-2-3.`DeleteHistoryTest`
            - [x] 0-5-2-4.`DeleteHistoryRepositoryTest`
            - [x] 0-5-2-5.`QuestionTest`
            - [x] 0-5-2-6.`QuestionRepositoryTest`
            - [x] 0-5-2-7.`UserTest`
            - [x] 0-5-2-8.`UserRepositoryTest`

Resolve: #134
See Also:

gregolee committed:
chore: java convention - google 적용

일부 코드 Convention 준수하도록 변경

Resolve:
See Also:

gregolee committed:
docs: 학습 내용 정리

- 학습 내용 정리
   - 영속성 전이 : CASCADE
   - 래거시 코드 리팩터링

Resolve:
See Also:

gregolee committed:
docs: TODO List 수정

- TODO List 수정
- [ ] 4.구현
    - [ ] 4-1.TDD
        - [ ] 4-1-1.일급컬렉션 적용
            - [ ] 4-1-1-1.`AnswerGroup`
        - [ ] 4-1-2.CASCADE 적용
            - [ ] 4-1-2-1.`Question` : `Answer` 양방향 연관 관계 매핑
            - [ ] 4-1-2-2.`Question` : CASCADE 적용
        - [ ] 4-1-3.validation
            - [ ] 4-1-3-1.`Question` : 삭제 유효성 검사
                - [ ] 4-1-3-1-1.`Question` : 사용자 동일 유효성 검사
                - [ ] 4-1-3-1-2.`Question` : 답변 없음 유효성 검사
                - [ ] 4-1-3-1-3.`Question` : 전체 답변 사용자 동일 유효성 검사
            - [ ] 4-1-3-2.`Answer` : 삭제 유효성 검사
                - [ ] 4-1-3-2-1.`Answer` : 사용자 동일 유효성 검사
            - [ ] 4-1-3-3.`AnswerGroup` : 삭제 유효성 검사
                - [ ] 4-1-3-3-1.`AnswerGroup` : 사용자 동일 유효성 검사
                - [ ] 4-1-3-3-2.`AnswerGroup` : 답변 없음 유효성 검사
                - [ ] 4-1-3-3-3.`AnswerGroup` : 전체 답변 사용자 동일 유효성 검사
        - [ ] 4-1-4.`repository.delete()` -> soft delete로 변경
            - [ ] 4-1-4-1.`Question` : 질문 삭제
            - [ ] 4-1-4-2.`Answer` : 답변 삭제
            - [ ] 4-1-4-3.`AnswerGroup` : 답변 그룹 삭제
        - [ ] 4-1-5.`DeleteHistory` : 삭제 완료 후 로깅 처리
    - [ ] 4-2.추가 리팩터링
        - [ ] 4-2-1.
- [ ] 5.테스트
    - [ ] 5-1.Gradle build Success 확인
    - [ ] 5-2.Google Java Style 적용 (indent : 2 -> 4 spaces)
    - [ ] 5-3.요구사항 조건들 충족했는지 확인
        - [ ] 5-3-1.핵심 단위 로직 테스트
        - [ ] 5-3-2.< 2 Depth 규칙
        - [ ] 5-3-3.함수(또는 메서드) 10라인 이하
        - [ ] 5-3-4.else 예약어 금지
        - [ ] 5-3-5.Code값들은 `enum`을 활용한다.
        - [ ] 5-3-6.모든 원시값과 문자열을 포장한다.
        - [ ] 5-3-7.일급콜렉션 활용
        - [ ] 5-3-8.축약 금지
        - [ ] 5-3-9.예외 처리
        - [ ] 5-3-10.3항 연산자 금지

Resolve:
See Also:

gregolee committed:
docs: ATDD 초안 작성

- ATDD 초안 작성
   - 2.3. 인수 조건 (ATDD)

Resolve:
See Also:

gregolee committed:
feat: 3단계 질문 삭제하기 리팩터링 구현 완료

- [ ] 4.구현
    - [x] 4-1.TDD
        - [x] 4-1-1.일급컬렉션 적용
            - [x] 4-1-1-1.`AnswerGroup`
            - [x] 4-1-1-2.`DeleteHistoryGroup`
        - [x] 4-1-2.CASCADE 적용
            - [x] 4-1-2-1.`AnswerGroup` : `Answer`-`Question` 양방향 연관 관계 매핑
            - [x] 4-1-2-2.`AnswerGroup` : CASCADE 적용
        - [x] 4-1-3.validation
            - [x] 4-1-3-1.`Question` : 삭제 유효성 검사
                - [x] 4-1-3-1-1.`Question` : 사용자 동일 유효성 검사
                - [x] 4-1-3-1-2.`Question` : 전체 답변 사용자 동일 유효성 검사
            - [x] 4-1-3-2.`Answer` : 삭제 유효성 검사
                - [x] 4-1-3-2-1.`Answer` : 사용자 동일 유효성 검사
            - [x] 4-1-3-3.`AnswerGroup` : 삭제 유효성 검사
                - [x] 4-1-3-3-1.`AnswerGroup` : 사용자 동일 유효성 검사
                - [x] 4-1-3-3-2.`AnswerGroup` : 전체 답변 사용자 동일 유효성 검사
        - [x] 4-1-4.`repository.delete()` -> soft delete로 변경
            - [x] 4-1-4-1.`Question` : 질문 삭제
            - [x] 4-1-4-2.`Answer` : 답변 삭제
            - [x] 4-1-4-3.`AnswerGroup` : 답변 그룹 삭제
        - [x] 4-1-5.`DeleteHistory` : 삭제 완료 후 로깅 처리
    - [ ] 4-2.추가 리팩터링
        - [ ] 4-2-1.

Resolve:
See Also:

gregolee committed:
refactor: 불필요한 메서드 제거

- [x] 4.구현
    - [x] 4-1.TDD
        - [x] 4-1-1.일급컬렉션 적용
            - [x] 4-1-1-1.`AnswerGroup`
            - [x] 4-1-1-2.`DeleteHistoryGroup`
        - [x] 4-1-2.CASCADE 적용
            - [x] 4-1-2-1.`AnswerGroup` : `Answer`-`Question` 양방향 연관 관계 매핑
            - [x] 4-1-2-2.`AnswerGroup` : CASCADE 적용
        - [x] 4-1-3.validation
            - [x] 4-1-3-1.`Question` : 삭제 유효성 검사
                - [x] 4-1-3-1-1.`Question` : 사용자 동일 유효성 검사
                - [x] 4-1-3-1-2.`Question` : 전체 답변 사용자 동일 유효성 검사
            - [x] 4-1-3-2.`Answer` : 삭제 유효성 검사
                - [x] 4-1-3-2-1.`Answer` : 사용자 동일 유효성 검사
            - [x] 4-1-3-3.`AnswerGroup` : 삭제 유효성 검사
                - [x] 4-1-3-3-1.`AnswerGroup` : 사용자 동일 유효성 검사
                - [x] 4-1-3-3-2.`AnswerGroup` : 전체 답변 사용자 동일 유효성 검사
        - [x] 4-1-4.`repository.delete()` -> soft delete로 변경
            - [x] 4-1-4-1.`Question` : 질문 삭제
            - [x] 4-1-4-2.`Answer` : 답변 삭제
            - [x] 4-1-4-3.`AnswerGroup` : 답변 그룹 삭제
        - [x] 4-1-5.`DeleteHistory` : 삭제 완료 후 로깅 처리
    - [x] 4-2.추가 리팩터링
        - [x] 4-2-1.불필요한 메서드 제거

Resolve:
See Also:

gregolee committed:
test: Build, CheckStyle 등 테스트

- [x] 5.테스트
    - [x] 5-1.Gradle build Success 확인
    - [x] 5-2.Google Java Style 적용 (indent : 2 -> 4 spaces)
    - [x] 5-3.요구사항 조건들 충족했는지 확인
        - [x] 5-3-1.핵심 단위 로직 테스트
        - [x] 5-3-2.< 2 Depth 규칙
        - [x] 5-3-3.함수(또는 메서드) 10라인 이하
        - [x] 5-3-4.else 예약어 금지
        - [x] 5-3-5.Code값들은 `enum`을 활용한다.
        - [x] 5-3-6.모든 원시값과 문자열을 포장한다.
        - [x] 5-3-7.일급콜렉션 활용
        - [x] 5-3-8.축약 금지
        - [x] 5-3-9.예외 처리
        - [x] 5-3-10.3항 연산자 금지

Resolve:
See Also:

gregolee committed:
chore: 인수인계 준비

- markdown 확인
- 최종 테스트 확인
- 코드리뷰 진행
- 피드백 요청
- 느낀점 및 배운점 작성

Resolve:
See Also:

gregolee committed:
chore: 강현구 리뷰어님께서 말씀하신 코멘트 정리

- [ ] 4-3.리뷰어님 코멘트 반영
        - [ ] 4-3-1.불필요한 원시값 포장 해제하기 : `Deleted`
        - [ ] 4-3-2.2중 적용된 소스 코드 수정 : `Answer.changeQuestion()`
        - [ ] 4-3-3.Checked Exception -> Unchecked Exception 적용
            - [ ] 4-3-3-1.`CannotDeleteException`
            - [ ] 4-3-3-2.`Answer.validateIsOwner()`
            - [ ] 4-3-3-3.`Question.validateCouldDelete()`
        - [ ] 4-3-4.생성자 -> 정적 팩토리 메서드
            - [ ] 4-3-4-1.`AnswerGroup.generateDeleteHIstoryAllOfAnswers()` 내 `new DeleteHistory`
        - [ ] 4-3-5.기능수정
            - [ ] 4-3-5-1.`Answer.delete()` : 상태 값 변경 -> 상태 값 변경 + 삭제한 이력을 리턴
        - [ ] 4-3-6.래핑클래스 -> 원시값으로 변경
            - [ ] 4-3-6-1.`Question.deleted`
        - [ ] 4-3-7.필드의 기본값 설정과 초기화를 혼동하지 말자.
            - [ ] 4-3-7-1.`Question.answers` : AnswerGroup.generate() -> 제거
        - [ ] 4-3-8.접근제어자 제대로 사용하기
            - [ ] 4-3-8-1.`Question.validateCouldDelete()`
            - [ ] 4-3-8-2.`Answer.validateIsOwner()`
            - [ ] 4-3-8-3.그 외에도 찾아보기
        - [ ] 4-3-9.Value Object Test 코드 작성

Resolve: #151
See Also:

gregolee committed:
refactor: 강현구 리뷰어님께서 알려주신 코멘트 수정(80%)

- [ ] 4-3.리뷰어님 코멘트 반영
        - [x] 4-3-1.불필요한 원시값 포장 해제하기 : `Deleted`
        - [x] 4-3-2.2중 적용된 소스 코드 수정 : `Answer.changeQuestion()`
        - [x] 4-3-3.Checked Exception -> Unchecked Exception 적용
            - [x] 4-3-3-1.`CannotDeleteException`
            - [x] 4-3-3-2.`Answer.validateIsOwner()`
            - [x] 4-3-3-3.`Question.validateCouldDelete()`
        - [x] 4-3-4.생성자 -> 정적 팩토리 메서드
            - [x] 4-3-4-1.`AnswerGroup.generateDeleteHIstoryAllOfAnswers()` 내 `new DeleteHistory`
        - [x] 4-3-5.기능수정
            - [x] 4-3-5-1.`Answer.delete()` : 상태 값 변경 -> 상태 값 변경 + 삭제한 이력을 리턴
        - [x] 4-3-6.래핑클래스 -> 원시값으로 변경
            - [x] 4-3-6-1.`Question.deleted`
        - [x] 4-3-7.필드의 기본값 설정과 초기화를 혼동하지 말자.
            - [x] 4-3-7-1.`Question.answers` : AnswerGroup.generate() -> 제거
        - [x] 4-3-8.접근제어자 제대로 사용하기
            - [x] 4-3-8-1.`Question.validateCouldDelete()`
            - [x] 4-3-8-2.`Answer.validateIsOwner()`
            - [x] 4-3-8-3.그 외에도 찾아보기
        - [ ] 4-3-9.Value Object Test 코드 작성

Resolve: #151
See Also:

gregolee committed:
refactor: package 재구성

domain
-> domain.aggregate
-> domain.code
-> domain.entity
-> domain.repository
-> domain.vo

Resolve:
See Also:

gregolee committed:
refactor: 강현구 리뷰어님께서 알려주신 코멘트 수정(100%)

- [x] 4-3.리뷰어님 코멘트 반영
        - [x] 4-3-1.불필요한 원시값 포장 해제하기 : `Deleted`
        - [x] 4-3-2.2중 적용된 소스 코드 수정 : `Answer.changeQuestion()`
        - [x] 4-3-3.Checked Exception -> Unchecked Exception 적용
            - [x] 4-3-3-1.`CannotDeleteException`
            - [x] 4-3-3-2.`Answer.validateIsOwner()`
            - [x] 4-3-3-3.`Question.validateCouldDelete()`
        - [x] 4-3-4.생성자 -> 정적 팩토리 메서드
            - [x] 4-3-4-1.`AnswerGroup.generateDeleteHIstoryAllOfAnswers()` 내 `new DeleteHistory`
        - [x] 4-3-5.기능수정
            - [x] 4-3-5-1.`Answer.delete()` : 상태 값 변경 -> 상태 값 변경 + 삭제한 이력을 리턴
        - [x] 4-3-6.래핑클래스 -> 원시값으로 변경
            - [x] 4-3-6-1.`Question.deleted`
        - [x] 4-3-7.필드의 기본값 설정과 초기화를 혼동하지 말자.
            - [x] 4-3-7-1.`Question.answers` : AnswerGroup.generate() -> 제거
        - [x] 4-3-8.접근제어자 제대로 사용하기
            - [x] 4-3-8-1.`Question.validateCouldDelete()`
            - [x] 4-3-8-2.`Answer.validateIsOwner()`
            - [x] 4-3-8-3.그 외에도 찾아보기
        - [x] 4-3-9.Value Object Test 코드 작성

Resolve: #151
See Also:

gregolee committed:
test: Build, CheckStyle 등 테스트

- [x] 5.테스트
    - [x] 5-1.Gradle build Success 확인
    - [x] 5-2.Google Java Style 적용 (indent : 2 -> 4 spaces)
    - [x] 5-3.요구사항 조건들 충족했는지 확인
        - [x] 5-3-1.핵심 단위 로직 테스트
        - [x] 5-3-2.< 2 Depth 규칙
        - [x] 5-3-3.함수(또는 메서드) 10라인 이하
        - [x] 5-3-4.else 예약어 금지
        - [x] 5-3-5.Code값들은 `enum`을 활용한다.
        - [x] 5-3-6.모든 원시값과 문자열을 포장한다.
        - [x] 5-3-7.일급콜렉션 활용
        - [x] 5-3-8.축약 금지
        - [x] 5-3-9.예외 처리
        - [x] 5-3-10.3항 연산자 금지

Resolve:
See Also:

gregolee committed:
chore: 인수인계 준비

- markdown 확인
- 최종 테스트 확인
- 코드리뷰 진행

Resolve:
See Also:

kang-hyungu commented on src/main/java/qna/domain/aggregate/AnswerGroup.java:
```diff
@@ -39,23 +40,13 @@ public boolean isEmpty() {
-		for (Answer answer : answers) {
-			answer.validateIsOwner(loginUser);
-		}
+	public void checkAnswersAreSameWithUserAndWriter(User loginUser) {
```
답변 삭제할 때  클라이언트가 권한 체크하는 책임을 갖게 하는 것보다 AnswerGroup 내부에서 하는게 낫지 않을까요?

gregolee replied:
리뷰어님과 DM으로 대화를 나누었습니다
대화를 나누면서 **'객체지향' 을 제대로 모르고 있구나** 라는 생각이 들었습니다.

- Q. public 으로 유효성 검증 가능하도록 열어두는 것은 옳지 않은가요?
- A. 필요하다면 public으로 둬야겠죠. 하지만 지금 구현하신 내용을 보면 delete 메서드에서만 필요한 상태죠. 그리고 삭제할 때 본인만 삭제 가능하도록 강제해야 의도한 기능이 맞지 않을까요? 삭제 정책이 본인만 가능하다면 기능도 제한을 줘야하는데 지금처럼 해당 기능 밖에서 처리하도록 만들면 유효성 검증을 놓쳐서 다른 사람이 삭제할 가능성이 생기겠죠
- A. 객체 내부에서 관리하는 인스턴스 변수에 대한 처리는 클래스 안에서만 처리하도록 강제해야 캡슐화를 달성할 수 있습니다

패러다임을 바꾸도록 생각하는게 정말 어려웠습니다 ㅠㅠ
그럼에도 불구하고 바꾸도록 도와주셔서 감사합니다!!
객체지향에 대해서 제대로 생각할 수 있도록 도움될만한 서적 꼭 읽어보도록 하겠습니다!!

kang-hyungu commented on src/main/java/qna/domain/aggregate/AnswerGroup.java:
```diff
@@ -39,23 +40,13 @@ public boolean isEmpty() {
-	}
-
-	public List<DeleteHistory> generateDeleteHistoryAllOfAnswers() {
+	public List<DeleteHistory> deleteAll() {
```
```suggestion
	public DeleteHistoryGroup deleteAll(User loginUser) {
```

1. 삭제하면서 권한 체크하도록 만들면 어떨까요
2. 일급 컬렉션을 활용해보는게 좋겠네요

gregolee replied:
DM을 통해 말씀하신 내용을 이해했습니다 💡 💡 💡 
객체지향으로 사고하는 것이 정말 어려운 것이라는 것을 절실히 느꼈습니다 😢 😱 
말씀하신대로 반영하도록 하겠습니다!

kang-hyungu commented on src/main/java/qna/domain/entity/Answer.java:
```diff
@@ -10,24 +11,26 @@
 @Entity
 public class Answer extends BaseEntity {
 
+	private static final boolean NOT_DELETED = false;
```
true, false를 상수로 만들 필요가 있을까요?
변수명이 삭제여부이고 그 값이 true, false인걸로 충분히 의미가 통하지 않을까요?

gregolee replied:
`매직넘버`를 원시값 전체를 대상으로 하여 상수를 만드는 것으로 생각했었습니다 😢 
말씀하신대로도 충분히 의미가 전달되기에 해당 상수는 제거하고, true, false만으로 수정하도록 하겠습니다!

kang-hyungu commented on src/test/java/qna/domain/entity/AnswerTest.java:
```diff
@@ -8,28 +8,29 @@
 
 public class AnswerTest {
```
정상 케이스 외에 의도한 예외도 성공적으로 발생하는지 테스트로 검증해보는게 좋겠네요.
생성자도 유효성 체크 잘 동작하는지 같이 체크해보아요

gregolee replied:
일전에 테스트는 성공한 값만 하는 것이 아니라 경계값을 잘 따져야 한다는 점을 읽은 적이 있습니다!
실패한 것도 예상 범위로 상정하고 테스트를 해야한다는 점...기본에 충실하게 반영해보겠습니다!
감사합니다!! 🙇

kang-hyungu commented on src/test/java/qna/domain/entity/DeleteHistoryTest.java:
```diff
@@ -18,23 +17,21 @@
-		deleteHistoryOfQuestionWrittenByJavajigi = new DeleteHistory(ContentType.QUESTION,
-			questionWrittenByJavajigi.id(), questionWrittenByJavajigi.writer(),
-			LocalDateTime.now());
+		deleteHistoryOfQuestionWrittenByJavajigi = DeleteHistory.ofQuestion(
```
변수명이 너무 길어지네요.
어차피 변수에 타입이 명시되어 있으니 어떤 변수인지 의미만 부여해보면 어떨까요

```suggestion
		자바지기_질문_삭제_이력 = DeleteHistory.ofQuestion(
```

gregolee replied:
네이밍은 아직도 어려운 부분입니다 😢 
지금처럼 말씀하신대로
1. 타입은 명시가 되어 있으니 변수에는 타입을 명시하지 않도록 하고
2. 변수의 의미만 부여하여 읽기 좋게 해보자

라는 의미로 이해하겠습니다!
그리고 한글 변수명도 적극적으로 활용해 보겠습니다!!
👍 🙇

kang-hyungu commented on src/test/java/qna/domain/entity/DeleteHistoryTest.java:
```diff
@@ -1,15 +1,14 @@
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
 
 public class DeleteHistoryTest {
```
정적 팩터리 메서드가 의도한 컨텐츠 타입으로 생성되는지 테스트해봐도 좋겠네요

gregolee replied:
넵! 감사합니다! 👍 🙇 🥇 
테스트를 조그만 단위로 세분화하여 하는 예제의 의도까지!!!
말씀하신 의도를 잘 이해하고 적용해보도록 하겠습니다!

kang-hyungu commented on src/test/java/qna/domain/vo/ContentIdTest.java:
```diff
@@ -0,0 +1,27 @@
+	@DisplayName("generate() : null 금지")
+	@ParameterizedTest(name = "{index} - contentId:[{0}], exceptedNotThrownException:{1}")
+	@CsvSource(value = {";false", "0;true", "100;true"}, delimiter = ';')
+	void generate(Long contentId, boolean exceptedNotThrownException) {
```
클린코드 25p

의미 있게 구분하라
컴파일러나 인터프리터만 통과하려는 생각으로 코드를 구현하는 프로그래머는 스스로 문제를 일으킨다.
예를 들어, 동일한 범위 안에서 다른 두 개념에 같은 이름을 사용하지 못한다.
...

컴파일러를 통과할지라도 연속된 숫자를 덧붙이거나 불용어(noise word)를 추가하는 방식은 적절하지 못하다.
이름이 달라야 한다면 의미도 달라져야 한다.

연속적인 숫자를 덧붙인 이름(a1, a2 ... aN)은 의도적인 이름과 정반대다.
이런 이름은 그릇된 정보를 제공하는 이름도 아니며, 아무런 정보를 제공하지 못하는 이름일 뿐이다.
저자 의도가 전혀 드러나지 않는다.


클린코드 157p

테스트 코드는 실제 코드 못지 않게 중요하다. 테스트 코드는 이류 시민이 아니다.
테스트 코드는 사고와 설계와 주의가 필요하다.
실제 코드 못지 않게 깨끗하게 짜야한다.

테스트는 유연성, 유지보수성, 재사용성을 제공한다. 테스트 케이스가 있으면 변경이 쉬워지기 때문이다.
따라서 테스트 코드가 지저분하면 코드를 변경하는 능력이 떨어지며 코드 구조를 개선하는 능력도 떨어진다.
테스트 코드가 지저분할수록 실제 코드도 지저분해진다.
결국 테스트 코드를 잃어버리고 실제 코드도 망가진다.

gregolee replied:
이 리뷰를 다시 보게 되어 참 씁쓸했습니다. 😢 
그만큼 제가 테스트 코드를 프로덕션 코드만큼 중요하게 생각하지 않았다는 반증이었습니다.💢 
다시 한 번 상기 시켜주심에 감사드립니다! 
다음에는 이 리뷰를 보지 않도록 잘 짜보겠습니다!!! 🔥 🔥 🔥 🔥 🔥

kang-hyungu commented on src/test/java/qna/domain/vo/ContentIdTest.java:
```diff
@@ -0,0 +1,27 @@
+	@DisplayName("generate() : null 금지")
+	@ParameterizedTest(name = "{index} - contentId:[{0}], exceptedNotThrownException:{1}")
+	@CsvSource(value = {";false", "0;true", "100;true"}, delimiter = ';')
+	void generate(Long contentId, boolean exceptedNotThrownException) {
```
모든 조건을 한 테스트에서 다 처리하기보다는 테스트 목적에 따라 테스트 메서드를 분리해주세요

테스트 코드에 if절이 존재할 이유가 없습니다.

gregolee replied:
리뷰어님과 반대로 생각했던 부분이었습니다.
**하나의 기능을 테스트하는 것을 하나의 메서드에서 처리하도록 한다.**고 생각했습니다.
하지만 곰곰이 생각해보면 그 테스트 코드를 이해하기 위해 다시 분석을 해야 한다.
즉, 코드를 한 번 읽어서 이해하기 어려워진다는 것임을 의미할 것으로 생각합니다.

말씀하신대로 테스트 메서드를 분리하겠습니다.
1. 같은 기능이라도 다른 Exception을 발생한다면, 그 발생하는 exception이 발생하는 조건의 테스트 코드를 작성한다.
2. 의도한 기능이 정상수행임에도 여러가지 결과를 도출한다면 각 결과에 맞춰서 테스트 코드를 작성한다.
3. if절은 활용하지 않는 것이 원칙이다.

해당 내용 명심하여 테스트 코드 다시 작성해보겠습니다!
감사합니다!!! 🙇 👍

kang-hyungu reviewed (changes requested):
재영님 피드백 반영 잘 하셨어요
조금 더 고쳐보면 좋을 부분에 피드백 남겼습니다.
수정해보시고 리뷰 재요청주세요.
궁금한 점은 코멘트 남기시거나 dm 주세요

gregolee commented:
이번 미션을 진행하면서 뭔가 해결되지 않은 듯한 답답함이 존재했는데,
제가 객체지향 패러다임이 갖춰지지 않았기 때문이었습니다.
또한 클린코드에서 말하는 테스트 코드를 관리하는 것에 대한 부족함을 다시 한 번 느낍니다
정말 반성하여 다음 리뷰 재요청에는 잘 반영하도록 해보겠습니다!

다음은 리뷰어님께서 말씀하신 내용을 반영할 정리본 입니다.
리뷰어님 정말 감사합니다! 👍 🙇 

    - [ ] 4-4.리뷰어님 코멘트 반영 2차
        - [ ] 4-4-1.답변 삭제 권한 체크는 AnswerGroup 내부 또는 Answer에서 하도록 변경
            - [ ] 4-4-1-1.`AnswerGroup.deleteAll()` -> `public DeleteHistoryGroup deleteAll(User loginUser)`
        - [ ] 4-4-2.불필요한 상수 적용 제거
            - [ ] 4-4-2-1."DELETED", "NOT_DELETED" -> `true`, `false`
        - [ ] 4-4-3.테스트 코드 리펙토링
            - [ ] 4-4-3-1.다양한 테스트 메서드 작성
                - [ ] 4-4-3-1-1.정상 케이스 외에 의도한 예외마다 발생하는 테스트 코드 작성
                - [ ] 4-4-3-1-2.정상 케이스임에도 다른 의도가 담긴 테스트 메서드로 분리
                - [ ] 4-4-3-1-3.정상 수행임에도 여러가지 결과를 도출한다면 각 결과에 맞춰서 테스트 코드 작성
                - [ ] 4-4-3-1-4.if절은 활용하지 않을 것
            - [ ] 4-4-3-2.네이밍
                - [ ] 4-4-3-2-1.타입은 명시하지 않도록 한다.
                - [ ] 4-4-3-2-2.변수의 의미만 부여한다.
                - [ ] 4-4-3-2-3.한글 변수명을 사용해보자.
            - [ ] 4-4-3-3.`DeleteHistoryTest` : 팩터리 메서드가 의도한 컨텐츠 타입으로 생성되는지 테스트

gregolee committed:
chore: 강현구 리뷰어님께서 말씀하신 두번째 코멘트 정리

    - [ ] 4-4.리뷰어님 코멘트 반영 2차
        - [ ] 4-4-1.답변 삭제 권한 체크는 AnswerGroup 내부 또는 Answer에서 하도록 변경
            - [ ] 4-4-1-1.`AnswerGroup.deleteAll()` -> `public DeleteHistoryGroup deleteAll(User loginUser)`
        - [ ] 4-4-2.불필요한 상수 적용 제거
            - [ ] 4-4-2-1."DELETED", "NOT_DELETED" -> `true`, `false`
        - [ ] 4-4-3.테스트 코드 리펙토링
            - [ ] 4-4-3-1.다양한 테스트 메서드 작성
                - [ ] 4-4-3-1-1.정상 케이스 외에 의도한 예외마다 발생하는 테스트 코드 작성
                - [ ] 4-4-3-1-2.정상 케이스임에도 다른 의도가 담긴 테스트 메서드로 분리
                - [ ] 4-4-3-1-3.정상 수행임에도 여러가지 결과를 도출한다면 각 결과에 맞춰서 테스트 코드 작성
                - [ ] 4-4-3-1-4.if절은 활용하지 않을 것
            - [ ] 4-4-3-2.네이밍
                - [ ] 4-4-3-2-1.타입은 명시하지 않도록 한다.
                - [ ] 4-4-3-2-2.변수의 의미만 부여한다.
                - [ ] 4-4-3-2-3.한글 변수명을 사용해보자.
            - [ ] 4-4-3-3.`DeleteHistoryTest` : 팩터리 메서드가 의도한 컨텐츠 타입으로 생성되는지 테스트

Resolve: #158
See Also:

gregolee committed:
refactor: 강현구 리뷰어님께서 알려주신 코멘트 수정(40%)

- [ ] 4-4.리뷰어님 코멘트 반영 2차
        - [x] 4-4-1.답변 삭제 권한 체크는 AnswerGroup 내부 또는 Answer에서 하도록 변경
            - [x] 4-4-1-1.`AnswerGroup.deleteAll()` -> `public DeleteHistoryGroup deleteAll(User loginUser)`
        - [x] 4-4-2.불필요한 상수 적용 제거
            - [x] 4-4-2-1."DELETED", "NOT_DELETED" -> `true`, `false`
        - [ ] 4-4-3.테스트 코드 리펙토링
            - [ ] 4-4-3-1.다양한 테스트 메서드 작성
                - [ ] 4-4-3-1-1.정상 케이스 외에 의도한 예외마다 발생하는 테스트 코드 작성
                - [ ] 4-4-3-1-2.정상 케이스임에도 다른 의도가 담긴 테스트 메서드로 분리
                - [ ] 4-4-3-1-3.정상 수행임에도 여러가지 결과를 도출한다면 각 결과에 맞춰서 테스트 코드 작성
                - [ ] 4-4-3-1-4.if절은 활용하지 않을 것
            - [ ] 4-4-3-2.네이밍
                - [ ] 4-4-3-2-1.타입은 명시하지 않도록 한다.
                - [ ] 4-4-3-2-2.변수의 의미만 부여한다.
                - [ ] 4-4-3-2-3.한글 변수명을 사용해보자.
            - [ ] 4-4-3-3.`DeleteHistoryTest` : 팩터리 메서드가 의도한 컨텐츠 타입으로 생성되는지 테스트

Resolve: #158
See Also:

gregolee committed:
refactor: 강현구 리뷰어님께서 알려주신 코멘트 수정(90%)

- [ ] 4-4.리뷰어님 코멘트 반영 2차
        - [x] 4-4-1.답변 삭제 권한 체크는 AnswerGroup 내부 또는 Answer에서 하도록 변경
            - [x] 4-4-1-1.`AnswerGroup.deleteAll()` -> `public DeleteHistoryGroup deleteAll(User loginUser)`
        - [x] 4-4-2.불필요한 상수 적용 제거
            - [x] 4-4-2-1."DELETED", "NOT_DELETED" -> `true`, `false`
        - [ ] 4-4-3.테스트 코드 리펙토링
            - [ ] 4-4-3-1.다양한 테스트 메서드 작성
                - [ ] 4-4-3-1-1.정상 케이스 외에 의도한 예외마다 발생하는 테스트 코드 작성
                - [ ] 4-4-3-1-2.정상 케이스임에도 다른 의도가 담긴 테스트 메서드로 분리
                - [ ] 4-4-3-1-3.정상 수행임에도 여러가지 결과를 도출한다면 각 결과에 맞춰서 테스트 코드 작성
                - [ ] 4-4-3-1-4.if절은 활용하지 않을 것
            - [ ] 4-4-3-2.네이밍
                - [ ] 4-4-3-2-1.타입은 명시하지 않도록 한다.
                - [ ] 4-4-3-2-2.변수의 의미만 부여한다.
                - [ ] 4-4-3-2-3.한글 변수명을 사용해보자.
            - [ ] 4-4-3-3.`DeleteHistoryTest` : 팩터리 메서드가 의도한 컨텐츠 타입으로 생성되는지 테스트

Resolve: #158
See Also:

gregolee committed:
refactor: 강현구 리뷰어님께서 알려주신 코멘트 수정(100%)

- [x] 4-4.리뷰어님 코멘트 반영 2차
        - [x] 4-4-1.답변 삭제 권한 체크는 AnswerGroup 내부 또는 Answer에서 하도록 변경
            - [x] 4-4-1-1.`AnswerGroup.deleteAll()` -> `public DeleteHistoryGroup deleteAll(User loginUser)`
        - [x] 4-4-2.불필요한 상수 적용 제거
            - [x] 4-4-2-1."DELETED", "NOT_DELETED" -> `true`, `false`
        - [x] 4-4-3.테스트 코드 리펙토링
            - [x] 4-4-3-1.다양한 테스트 메서드 작성
                - [x] 4-4-3-1-1.정상 케이스 외에 의도한 예외마다 발생하는 테스트 코드 작성
                - [x] 4-4-3-1-2.정상 케이스임에도 다른 의도가 담긴 테스트 메서드로 분리
                - [x] 4-4-3-1-3.정상 수행임에도 여러가지 결과를 도출한다면 각 결과에 맞춰서 테스트 코드 작성
                - [x] 4-4-3-1-4.if절은 활용하지 않을 것
            - [x] 4-4-3-2.네이밍
                - [x] 4-4-3-2-1.타입은 명시하지 않도록 한다.
                - [x] 4-4-3-2-2.변수의 의미만 부여한다.
                - [x] 4-4-3-2-3.한글 변수명을 사용해보자.
            - [x] 4-4-3-3.`DeleteHistoryTest` : 팩터리 메서드가 의도한 컨텐츠 타입으로 생성되는지 테스트

Resolve: #158
See Also:

gregolee committed:
test: Build, CheckStyle 등 테스트

- [x] 5.테스트
    - [x] 5-1.Gradle build Success 확인
    - [x] 5-2.Google Java Style 적용 (indent : 2 -> 4 spaces)
    - [x] 5-3.요구사항 조건들 충족했는지 확인
        - [x] 5-3-1.핵심 단위 로직 테스트
        - [x] 5-3-2.< 2 Depth 규칙
        - [x] 5-3-3.함수(또는 메서드) 10라인 이하
        - [x] 5-3-4.else 예약어 금지
        - [x] 5-3-5.Code값들은 `enum`을 활용한다.
        - [x] 5-3-6.모든 원시값과 문자열을 포장한다.
        - [x] 5-3-7.일급콜렉션 활용
        - [x] 5-3-8.축약 금지
        - [x] 5-3-9.예외 처리
        - [x] 5-3-10.3항 연산자 금지

Resolve:
See Also:

gregolee committed:
chore: 인수인계 준비

- markdown 확인
- 최종 테스트 확인
- 코드리뷰 진행

Resolve:
See Also:

gregolee commented:
테스트 코드를 일류시민으로 끌어올리도록 노력했습니다! 🔥 🔥 🔥 🔥 🔥 
이번에는 잘 통과할 수 있기를!!! 😄 
좋은 하루 되시길 바랍니다!

    - [x] 4-4.리뷰어님 코멘트 반영 2차
        - [x] 4-4-1.답변 삭제 권한 체크는 AnswerGroup 내부 또는 Answer에서 하도록 변경
            - [x] 4-4-1-1.`AnswerGroup.deleteAll()` -> `public DeleteHistoryGroup deleteAll(User loginUser)`
        - [x] 4-4-2.불필요한 상수 적용 제거
            - [x] 4-4-2-1."DELETED", "NOT_DELETED" -> `true`, `false`
        - [x] 4-4-3.테스트 코드 리펙토링
            - [x] 4-4-3-1.다양한 테스트 메서드 작성
                - [x] 4-4-3-1-1.정상 케이스 외에 의도한 예외마다 발생하는 테스트 코드 작성
                - [x] 4-4-3-1-2.정상 케이스임에도 다른 의도가 담긴 테스트 메서드로 분리
                - [x] 4-4-3-1-3.정상 수행임에도 여러가지 결과를 도출한다면 각 결과에 맞춰서 테스트 코드 작성
                - [x] 4-4-3-1-4.if절은 활용하지 않을 것
            - [x] 4-4-3-2.네이밍
                - [x] 4-4-3-2-1.타입은 명시하지 않도록 한다.
                - [x] 4-4-3-2-2.변수의 의미만 부여한다.
                - [x] 4-4-3-2-3.한글 변수명을 사용해보자.
            - [x] 4-4-3-3.`DeleteHistoryTest` : 팩터리 메서드가 의도한 컨텐츠 타입으로 생성되는지 테스트

kang-hyungu commented on src/main/java/qna/domain/aggregate/AnswerGroup.java:
```diff
@@ -39,24 +38,12 @@ public boolean isEmpty() {
-				answer.writer(),
-				LocalDateTime.now()))
-			.collect(Collectors.toList());
+	public DeleteHistoryGroup deleteAll(User loginUser) {
```
```java
	public DeleteHistoryGroup deleteAll(User loginUser) {
		return answers.stream()
			.map(answer -> answer.delete(loginUser))
			.collect(Collectors.collectingAndThen(Collectors.toList(), DeleteHistoryGroup::new));
	}
```

생성자에서 리스트로 받도록 처리하면 더 깔끔하게 작성할 수 있겠네요

1. generate라는 정적 메서드 팩토리에 너무 의존하고 있습니다. 기본 생성자가 있는데 따로 팩터리 메서드를 써야 할 이유가 있을까요?
2. 기본생성자만 사용하지 말고 생성자에 인자를 받도록 만들면 처음부터 의도한 값으로 구성된 객체를 만들 수 있습니다.

gregolee replied:
리뷰어님께서 말씀하신대로 더 깔끔한 코드가 되어서 너무 감사합니다!!
**무언가 코드가 추가될 것을 대비한다**는 생각으로 다소 불필요한 코드를 작성한 것 같습니다.

> 1. generate라는 정적 메서드 팩토리에 너무 의존하고 있습니다. 기본 생성자가 있는데 따로 팩터리 메서드를 써야 할 이유가 있을까요?
> 2. 기본생성자만 사용하지 말고 생성자에 인자를 받도록 만들면 처음부터 의도한 값으로 구성된 객체를 만들 수 있습니다.

의도가 있었다면 그에 맞는 코드를 작성하도록 고민해야 된다는 점은 잘 새기도록 하겠습니다!
리뷰 감사합니다! 🙇 👍

kang-hyungu commented on src/main/java/qna/domain/entity/Question.java:
```diff
@@ -91,39 +97,26 @@ public boolean isDeleted() {
-		this.deleted = true;
-		answerGroup.deleteAll();
-		updatedAtNow();
+	public DeleteHistoryGroup delete(User loginUser) {
```
```java
	public DeleteHistoryGroup delete(User loginUser) {
		DeleteHistoryGroup deleteHistoryGroup = answerGroup.deleteAll(loginUser);
		deleteHistoryGroup.add(deleteQuestion(loginUser));
		return deleteHistoryGroup;
	}
```
더 간결하게 만들 수도 있겠네요

gregolee replied:
절차지향적인 생각이 여실히 드러난 코드였습니다.
순서대로 작성해야 된다는 생각때문에 불필요한 코드를 만들어내었습니다 😢 
조금 더 나은 코드를 작성하게 해주심에 감사드립니다!! 👍 🙇

kang-hyungu reviewed (approved):
재영님 피드백 반영 잘해주셨어요 👍 
pr은 머지 처리하겠습니다. 수고하셨어요!
